crypto/signer_test.go

```diff
@@ -94,3 +96,110 @@ func TestSignLogRootSignerFails(t *testing.T) {
 	_, err = NewSHA256Signer(s).Sign(hash)
 	testonly.EnsureErrorContains(t, err, "signfail")
 }
+
+func TestSignLogRoot(t *testing.T) {
+	key, err := pem.UnmarshalPrivateKey(testonly.DemoPrivateKey, testonly.DemoPrivateKeyPass)
+	if err != nil {
+		t.Fatalf("Failed to open test key, err=%v", err)
+	}
+	signer := NewSHA256Signer(key)
+
+	for _, test := range []struct {
+		root trillian.SignedLogRoot
+	}{
+		{root: trillian.SignedLogRoot{TimestampNanos: 2267709, RootHash: []byte("Islington"), TreeSize: 2}},
+	} {
+		sig, err := signer.SignLogRoot(&test.root)
+		if err != nil {
+			t.Errorf("Failed to sign log root: %v", err)
+		}
+		if got := len(sig.Signature); got == 0 {
+			t.Errorf("len(sig): %v, want > 0", got)
+		}
+		if got, want := sig.HashAlgorithm, sigpb.DigitallySigned_SHA256; got != want {
+			t.Errorf("Hash alg incorrect, got %v expected %d", got, want)
+		}
+		if got, want := sig.SignatureAlgorithm, sigpb.DigitallySigned_ECDSA; got != want {
+			t.Errorf("Sig alg incorrect, got %v expected %v", got, want)
+		}
+		// Check that the signature is correct
+		obj, err := HashLogRoot(test.root)
+		if err != nil {
+			t.Errorf("HashLogRoot err: got %v want nil", err)
+		}
+
+		if err := Verify(key.Public(), obj, sig); err != nil {
+			t.Errorf("Verify(%v) failed: %v", test.root, err)
+		}
+	}
+}
+
+func TestSignLogRoot_SignerFails(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	key, err := pem.UnmarshalPrivateKey(testonly.DemoPrivateKey, testonly.DemoPrivateKeyPass)
+	if err != nil {
+		t.Fatalf("Failed to load public key: %v", err)
+	}
+
+	s := testonly.NewSignerWithErr(key, errors.New("signfail"))
```

Comment:
I don't think this test adds value, since it's just exercising a fake/testonly Signer instead of the "real" one. At best, this (plus other tests that exercise NewSignerWithErr) should be moved to a `testonly/signer_test.go` file, but I'd just remove them altogether.

Ditto for the other tests based on NewSignerWithErr in this file.

Reply:
The idea of `testonly/foo_test.go` seems very odd to me, unless you want to test structures that themselves are only used in testing.

These tests I think are to provide coverage of (one of) the error paths through the Sign*() funcs.
I'll remove the two I just added since the error path is identical in each case to the existing Sign() test and it's `TestSign_SignerFails()` func.